### PR TITLE
feat(cirq): Add greedy implementation of aligning 1-qubit, and 2-qubit gates optimizer

### PR DESCRIPTION
**Pull request by KevinVillela**
This is a follow-up to https://github.com/quantumlib/Cirq/pull/2772/ which I mailed over a month ago (sorry, I was planning on doing this in early March, and then Covid happened...). It uses a much simpler implementation, and by trying slightly different configurations it actually gets pretty good results.

### CODE REVIEW

**Comment by KevinVillela** [comment]
Almost :) The current method increases the size of that circuit from 8 to 11. The new method increases the size from 8 to 9.

I also added a few other tests where the new method outputs shorter circuits than the current method, PTAL.

**Comment by KevinVillela** [comment]
Friendly ping @Strilanc

**Comment by balopat** [comment]
Hi @KevinVillela - I pressed the update branch button to trigger the Github Action based CI tests - I hope you are okay with that.

**Comment by KevinVillela** [comment]
Yea no worries! Is this still something we want? If not, I can drop the PR.

**Comment by balopat** [comment]
I think we do want it, sorry for the slow reviews here. 
@mpharrigan @maffoo @Strilanc do you guys mind taking another look? 

I will as well, but I need some time to get up to speed with the details so I understand it deeply - from a cursory look it looks good with all the tests and conversations.  

We also have perf season currently - so this couple of weeks will be slow.

**Comment by KevinVillela** [comment]
Sure thing, thanks for the update!

File: cirq/optimizers/stratify.py [hunk 1]
```diff
@@ -11,19 +11,23 @@
...
+from typing import TYPE_CHECKING, Type, Callable, Union, Iterable, Set
 
 from cirq import ops, circuits
 
 if TYPE_CHECKING:
     import cirq
 
+# Any of the possible operation categories that we can stratify on.
+Category = Union['cirq.Gate', 'cirq.Operation', Type['cirq.Gate'],
+                 Type['cirq.Operation'], Callable[['cirq.Operation'], bool]]
```
Comment (balopat) [inline cirq/optimizers/stratify.py:24]:
```suggestion
                 Type['cirq.Operation'], 'Classifier']
```

File: cirq/optimizers/stratify.py [hunk 2]
```diff
@@ -54,55 +58,96 @@ def stratified_circuit(circuit: 'cirq.Circuit', *, categories: Iterable[
...
-        return lambda op: op != classifier
-    elif isinstance(classifier, type) and issubclass(classifier, ops.Gate):
-        return lambda op: not isinstance(op.gate, classifier)
-    elif isinstance(classifier, type) and issubclass(classifier, ops.Operation):
-        return lambda op: not isinstance(op, classifier)
-    elif callable(classifier):
-        return lambda op: not classifier(op)
+    # Normalize categories into classifier functions.
+    classifiers = [
+        _category_to_classifier(classifier) for classifier in categories
```
Comment (balopat) [inline cirq/optimizers/stratify.py:63]:
```suggestion
        _category_to_classifier(category) for category in categories
```

File: cirq/optimizers/stratify.py [hunk 3]
```diff
@@ -54,55 +58,96 @@ def stratified_circuit(circuit: 'cirq.Circuit', *, categories: Iterable[
...
+    # Make the classifiers exhaustive by adding an "everything else" bucket.
+    and_the_rest = lambda op: all(
+        not classifier(op) for classifier in classifiers)
+    classifiers_and_the_rest = [*classifiers, and_the_rest]
+
+    # Try the algorithm with each permutation of the classifiers.
+    classifiers_permutations = list(
+        itertools.permutations(classifiers_and_the_rest))
+    solutions = []
+    for c in classifiers_permutations:
```
Comment (balopat) [inline cirq/optimizers/stratify.py:74]: You can merge the two for loops.

**Review by balopat — APPROVED** [review]
This looks good to me except some nits.
Thank you for your patience on this!

**Comment by KevinVillela** [comment]
Nice, thanks for picking this back up and merging!

**Pull request merged**